File: tests/machine/x/clj/group_by_left_join.clj
(ns main)

(defrecord _Group [key Items])

(defn _group_by [src keyfn]
  (let [groups (transient {})
        order (transient [])]
    (doseq [it src]
      (let [k (keyfn it)
            ks (str k)
            g (get groups ks)]
        (if g
          (assoc! groups ks (assoc g :Items (conj (:Items g) it)))
          (do
            (assoc! groups ks (_Group. k [it]))
            (conj! order ks))))
    (let [g (persistent! groups)
          o (persistent! order)]
      (mapv #(get g %) o))) )

(defn _query [src joins opts]
  (let [items (atom (mapv vector src))]
    (doseq [j joins]
      (let [joined (atom [])]
        (if (and (:leftKey j) (:rightKey j))
          (let [idx (group-by (:rightKey j) (:items j))
                seen (atom #{})]
            (doseq [left @items]
              (let [k ((:leftKey j) left)
                    rs (get idx k)]
                (if rs
                  (do
                    (swap! seen conj k)
                    (doseq [right rs]
                      (swap! joined conj (conj left right))))
                  (when (:left j)
                    (swap! joined conj (conj left nil)))))
            (when (:right j)
              (doseq [right (:items j)]
                (when-not (contains? @seen ((:rightKey j) right))
                  (swap! joined conj (vec (concat (repeat (count (first (or @items []))) nil) [right]))))))
            (reset! items @joined))
          (cond
            (and (:right j) (:left j))
              (let [matched (boolean-array (count (:items j)))]
                (doseq [left @items]
                  (let [m (atom false)]
                    (doseq [[ri right] (map-indexed vector (:items j))]
                      (let [keep (if-let [f (:on j)]
                                   (apply f (conj left right))
                                   true)]
                        (when keep
                          (reset! m true)
                          (aset matched ri true)
                          (swap! joined conj (conj left right))))
                    (when-not @m
                      (swap! joined conj (conj left nil))))
                (doseq [[ri right] (map-indexed vector (:items j))]
                  (when-not (aget matched ri)
                    (swap! joined conj (vec (concat (repeat (count (first (or @items []))) nil) [right])))))
              (reset! items @joined)
            (:right j)
              (do
                (doseq [right (:items j)]
                  (let [m (atom false)]
                    (doseq [left @items]
                      (let [keep (if-let [f (:on j)]
                                   (apply f (conj left right))
                                   true)]
                        (when keep
                          (reset! m true)
                          (swap! joined conj (conj left right))))
                    (when-not @m
                      (swap! joined conj (vec (concat (repeat (count (first (or @items []))) nil) [right])))))
                (reset! items @joined))
            :else
              (do
                (doseq [left @items]
                  (let [m (atom false)]
                    (doseq [right (:items j)]
                      (let [keep (if-let [f (:on j)]
                                   (apply f (conj left right))
                                   true)]
                        (when keep
                          (reset! m true)
                          (swap! joined conj (conj left right))))
                    (when (and (:left j) (not @m))
                      (swap! joined conj (conj left nil))))
                (reset! items @joined)))))
    (let [it @items
          it (if-let [w (:where opts)] (vec (filter #(apply w %) it)) it)
          it (if-let [sk (:sortKey opts)]
               (vec (sort-by #(let [k (apply sk %)] (_sort_key k)) it))
               it)
          it (if (contains? opts :skip) (vec (drop (:skip opts) it)) it)
          it (if (contains? opts :take) (vec (take (:take opts) it)) it)]
      (mapv #(apply (:select opts) %) it))))))))))))
(declare customers orders stats)

(defn -main []
  (def customers [{:id 1 :name "Alice"} {:id 2 :name "Bob"} {:id 3 :name "Charlie"}]) ;; list of 
  (def orders [{:id 100 :customerId 1} {:id 101 :customerId 1} {:id 102 :customerId 2}]) ;; list of 
  (def stats (let [_src customers
      _rows (_query _src [
        {:items orders :leftKey (fn [c] (:customerId o)) :rightKey (fn [o] (:id c)) :left true}
      ] { :select (fn [c o] [c o]) })
      _groups (_group_by _rows (fn [c o] (:name c)))
      ]
  (vec (map (fn [g] {:name (:key g) :count (count (vec (->> (for [r (:Items g) :when (:o r)] r))))}) _groups)))) ;; list of 
  (println "--- Group Left Join ---")
  (loop [_tmp0 (seq stats)]
    (when _tmp0
      (let [s (clojure.core/first _tmp0)]
        (let [r (try
          (println (:name s) "orders:" (:count s))
          :next
        (catch clojure.lang.ExceptionInfo e
          (cond
            (= (.getMessage e) "continue") :next
            (= (.getMessage e) "break") :break
            :else (throw e))
          )
        )]
      (cond
        (= r :break) nil
        :else (recur (next _tmp0))
      )
    )
  )
)
)
)

(-main)
)
)

File: tests/transpiler/x/php/group_by_left_join.php
<?php
$customers = [["id" => 1, "name" => "Alice"], ["id" => 2, "name" => "Bob"], ["id" => 3, "name" => "Charlie"]];
$orders = [["id" => 100, "customerId" => 1], ["id" => 101, "customerId" => 1], ["id" => 102, "customerId" => 2]];
$stats = (function() use ($customers, $orders) {
  $groups = [];
  foreach ($customers as $c) {
    $matched = false;
    foreach ($orders as $o) {
      if (!($o["customerId"] == $c["id"])) continue;
      $matched = true;
      $key = $c["name"];
      if (!array_key_exists($key, $groups)) {
        $groups[$key] = ['key' => $key, 'items' => []];
      }
      $groups[$key]['items'][] = ['c' => $c, 'o' => $o];
    }
    if (!$matched) {
      $o = null;
      $key = $c["name"];
      if (!array_key_exists($key, $groups)) {
        $groups[$key] = ['key' => $key, 'items' => []];
      }
      $groups[$key]['items'][] = ['c' => $c, 'o' => $o];
    }
  }
  $result = [];
  foreach ($groups as $g) {
      $result[] = ["name" => $g["key"], "count" => count((function() use ($c, $o, $g) {
  $result = [];
  foreach ($g["items"] as $r) {
    if ($r["o"]) {
      $result[] = $r;
    }
  }
  return $result;
})())];
  }
  return $result;
})();
echo rtrim("--- Group Left Join ---"), PHP_EOL;
foreach ($stats as $s) {
  echo rtrim((is_float($s["name"]) ? sprintf("%.15f", $s["name"]) : $s["name"]) . " " . "orders:" . " " . (is_float($s["count"]) ? sprintf("%.15f", $s["count"]) : $s["count"])), PHP_EOL;
}
?>
